## 1.Importing libraries

In [1]:
import os

import pandas as pd

import numpy as np

from sklearn.model_selection import train_test_split

## 2. Reading the data

In [2]:
project_dir = r"D:\Projects\Flight_Price_Prediction"
data_dir = "data"

In [3]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(project_dir,data_dir,file_name)
    return pd.read_csv(file_path)

In [4]:
flights = get_data("flight_price")
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

## 3. Premaliary Analysis

#### 3.1 Checking Data Type

In [5]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


#### 3.2 Check for Duplicates

In [6]:
flights.duplicated().sum()

np.int64(220)

In [7]:
(
 flights
 .loc[flights.duplicated()]
 .sort_values(['Airline','Date_of_Journey','Source','Destination'])
)

Airline Date_of_Journey    Source Destination  \
9848    Air India      01/03/2019  Banglore   New Delhi   
8168    Air India      03/03/2019  Banglore   New Delhi   
9913    Air India       1/04/2019   Kolkata    Banglore   
5042    Air India       1/05/2019   Kolkata    Banglore   
6377    Air India       1/05/2019   Kolkata    Banglore   
...           ...             ...       ...         ...   
9216  Jet Airways       9/06/2019     Delhi      Cochin   
7502     SpiceJet      03/03/2019  Banglore   New Delhi   
2870     SpiceJet      24/03/2019  Banglore   New Delhi   
3711     SpiceJet      24/03/2019  Banglore   New Delhi   
5470      Vistara      24/03/2019  Banglore   New Delhi   

                      Route Dep_Time  Arrival_Time Duration Total_Stops  \
9848  BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops   
8168              BLR → DEL    21:10         23:55   2h 45m    non-stop   
9913  CCU → DEL → COK → BLR    10:00  01:20 02 Apr  15h 20m     2 stops   
5042  CCU → DEL → COK → BLR    10:00  13:45 02 May  27h 45m     2 stops   
6377  CCU → DEL → COK → BLR    10:00  13:45 02 May  27h 45m     2 stops   
...                     ...      ...           ...      ...         ...   
9216  DEL → IDR → BOM → COK    05:30  04:25 10 Jun  22h 55m     2 stops   
7502              BLR → DEL    20:30         23:20   2h 50m    non-stop   
2870              BLR → DEL    05:45         08:35   2h 50m    non-stop   
3711              BLR → DEL    20:30         23:20   2h 50m    non-stop   
5470              BLR → DEL    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
9848                       No info  17135  
8168                       No info   7591  
9913                       No info  10408  
5042                       No info  15164  
6377                       No info  15164  
...                            ...    ...  
9216                       No info  13292  
7502                       No info   6860  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
5470                       No info   5403  

[220 rows x 11 columns]

In [8]:
flights['Source'].unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [9]:
flights['Destination'].unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

#### 3.3 Observations

--  Datatype of "Date_of_Journey" , 'Dep_Time', 'Arrival_Time' should be changed from object to datetime.
-- "Duration" , "Total_stops" contain mixed dtypes. it should be numeric 
--   There are total 220 duplicates values.
--   'Airline' feature contain data redundency due to error in collecting



### 4. Detailed Analysis:

#### Airline

In [10]:
flights['Airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [11]:
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .unique()
)

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

In [12]:
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
     .str.title()  # it capitalize first character
)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

##### Date_of_Journey

###### --  Datatype of "Date_of_Journey" should be changed from object to datetime.

In [13]:
pd.to_datetime(flights['Date_of_Journey'],dayfirst = True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

##### Dep_Time

In [15]:
flights['Dep_Time']

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [16]:
# checking that is it contain Dep_Time "a:b" format ot not 

In [17]:
 flights['Dep_Time'].str.contains("[^0-9:]")

0        False
1        False
2        False
3        False
4        False
         ...  
10678    False
10679    False
10680    False
10681    False
10682    False
Name: Dep_Time, Length: 10683, dtype: bool

In [18]:
(
    flights
    .Dep_Time
    .str.contains("[^0-9:]")
)

0        False
1        False
2        False
3        False
4        False
         ...  
10678    False
10679    False
10680    False
10681    False
10682    False
Name: Dep_Time, Length: 10683, dtype: bool

In [19]:
flights.head(5)

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

conclusion :
 it means it contain  all Dep_Time in the correct format

#### Arrival_Time

In [20]:
flights.Arrival_Time.str.split(" ",n=1).str.get(1).unique()

array(['22 Mar', nan, '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

#### Duration:

In [21]:
(
    flights
    .Duration
    # .str.split(" ", expand = True)
    # .set_axis(["hour","minute"],axis = 1)
    .loc[lambda ser : ~ser.str.contains("m")]
    .unique()
)


array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

seems valid in hour . let's heckout for the minute part

In [22]:
(
    flights
    .Duration
    # .str.split(" ", expand = True)
    # .set_axis(["hour","minute"],axis = 1)
    .loc[lambda ser : ~ser.str.contains("h")]
    .unique()
)


array(['5m'], dtype=object)

duration can't we 5 Minutes , Suspecious !

In [23]:
(
    flights
    .Duration
    # .str.split(" ", expand = True)
    # .set_axis(["hour","minute"],axis = 1)
    .loc[lambda ser : ~ser.str.contains("h")]
    
)


6474    5m
Name: Duration, dtype: object

index 6474 has suspecious value !!

Let's remove it form flights dataset

In [24]:
(
    flights
    .Duration
    .drop(index = 6474)
    .str.split(" ", expand = True)
    .set_axis(["hour","minute"],axis = 1)
    .assign(
        hour = lambda df_ : (
            df_
            .hour
            .str.replace("h","")
            .astype(int)
            .mul(60)
        ),

        minute = lambda df_ :(
            df_
            .minute
            .str.replace("m","")
            .fillna("0")
            .astype(int)
        )
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
)


duration_minutes
0                   170
1                   445
2                  1140
3                   325
4                   285
...                 ...
10678               150
10679               155
10680               180
10681               160
10682               500

[10682 rows x 1 columns]

##### total_stops

In [25]:
flights.head(5)

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [26]:
(
    flights
    .Total_Stops
    .replace("non-stop","0")
    .str.replace("stops?","",regex= True)
    .pipe(lambda ser:pd.to_numeric(ser))
    .fillna(0)
    .unique()
)

array([0., 2., 1., 3., 4.])

#### Additional_info

In [27]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

### 5.Cleaning Operations:

In [28]:
def convert_into_min(ser):
    return (
         ser
        .str.split(" ", expand = True)
        .set_axis(["hour","minute"],axis = 1)
        .assign(
            hour = lambda df_ : (
                df_
                .hour
                .str.replace("h","")
                .astype(int)
                .mul(60)
            ),
    
            minute = lambda df_ :(
                df_
                .minute
                .str.replace("m","")
                .fillna("0")
                .astype(int)
            )
        )
        .sum(axis=1)
    )

In [29]:
def clean_data(df):
    return (
        df
            .drop(index = 6474)
            .drop_duplicates()
    		.assign(**{
    			col: df[col].str.strip()
    			for col in df.select_dtypes(include="O").columns
    		})
            .rename(columns = str.lower)
            .assign(airline = lambda df_ : (
                df_
                .airline
                .str.replace(" Premium economy","")
                .str.replace(" Business","")
                .str.title()
                ),
                date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey, dayfirst=True),
                dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
    			arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
                duration = lambda df_: df_.duration.pipe(convert_into_min),
                total_stops = lambda df_ : (
                     df_
                     .total_stops
                    .replace("non-stop","0")
                    .str.replace("stops?","",regex= True)
                    .pipe(lambda ser:pd.to_numeric(ser))
                )	
            )
            .drop(columns="route")
    )

In [30]:
flights_cleaned = clean_data(flights)
flights_cleaned

C:\Users\Ravindra\AppData\Local\Temp\ipykernel_24812\1983535961.py:19: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\Ravindra\AppData\Local\Temp\ipykernel_24812\1983535961.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No info   3897  
1         13:15:00       445          2.0         No info   7662  
2         04:25:00      1140          2.0         No info  13882  
3         23:30:00       325          1.0         No info   6218  
4         21:35:00       285          1.0         No info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No info   4107  
10679     23:20:00       155          0.0         No info   4145  
10680     11:20:00       180          0.0         No info   7229  
10681     14:10:00       160          0.0         No info  12648  
10682     19:15:00       500          2.0         No info  11753  

[10462 rows x 10 columns]

In [31]:
flights_cleaned['airline'].unique()

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

In [32]:
flights_cleaned['date_of_journey']

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: date_of_journey, Length: 10462, dtype: datetime64[ns]

### Splitting the Data

In [33]:
flights_final =  flights_cleaned.sample(1000)
flights_final

airline date_of_journey    source destination  dep_time  \
8964  Jet Airways      2019-03-12  Banglore   New Delhi  21:25:00   
45       Spicejet      2019-03-24  Banglore   New Delhi  05:45:00   
4376  Jet Airways      2019-05-12   Kolkata    Banglore  18:55:00   
739   Jet Airways      2019-06-09   Kolkata    Banglore  21:10:00   
3891       Indigo      2019-03-01    Mumbai   Hyderabad  16:55:00   
...           ...             ...       ...         ...       ...   
1261  Jet Airways      2019-05-27     Delhi      Cochin  15:00:00   
9428      Vistara      2019-06-24     Delhi      Cochin  06:00:00   
4427  Jet Airways      2019-04-01   Kolkata    Banglore  16:30:00   
7903    Air India      2019-06-09   Kolkata    Banglore  09:25:00   
9823       Indigo      2019-05-09   Kolkata    Banglore  06:20:00   

     arrival_time  duration  total_stops               additional_info  price  
8964     08:15:00       650          1.0                       No info  12547  
45       08:35:00       170          0.0  No check-in baggage included   3527  
4376     12:00:00      1025          1.0   In-flight meal not included   9663  
739      16:20:00      1150          1.0   In-flight meal not included  10844  
3891     18:25:00        90          0.0                       No info   8853  
...           ...       ...          ...                           ...    ...  
1261     12:35:00      1295          1.0   In-flight meal not included  12898  
9428     09:10:00       190          0.0                       No info   6216  
4427     22:35:00       365          1.0   In-flight meal not included   7064  
7903     07:55:00      1350          1.0                       No info   8366  
9823     10:20:00       240          1.0                       No info   4226  

[1000 rows x 10 columns]

In [34]:
X = flights_final.drop(columns="price")
y = flights_final.price.copy()


In [35]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


 ### Export the Subsets

In [36]:
def export_data(X, y, name):
	file_name = f"{name}.csv"
	file_path = os.path.join(project_dir, data_dir, file_name)

	X.join(y).to_csv(file_path, index=False)

	return pd.read_csv(file_path).head()

In [37]:
export_data(X_train, y_train, "train")

airline date_of_journey    source destination  dep_time arrival_time  \
0       Indigo      2019-06-01     Delhi      Cochin  17:10:00     01:30:00   
1    Air India      2019-03-27     Delhi      Cochin  13:30:00     08:15:00   
2  Jet Airways      2019-03-06    Mumbai   Hyderabad  07:10:00     08:35:00   
3  Jet Airways      2019-06-12   Kolkata    Banglore  14:05:00     09:20:00   
4    Air India      2019-03-24  Banglore   New Delhi  21:05:00     23:55:00   

   duration  total_stops              additional_info  price  
0       500          1.0                      No info   6628  
1      1125          1.0                      No info   6681  
2        85          0.0                      No info   9536  
3      1155          1.0  In-flight meal not included   8529  
4       170          0.0                      No info   4441

In [38]:
export_data(X_val, y_val, "val")

airline date_of_journey    source destination  dep_time  \
0        Jet Airways      2019-06-06   Kolkata    Banglore  17:00:00   
1        Jet Airways      2019-06-27     Delhi      Cochin  20:00:00   
2        Jet Airways      2019-06-12   Kolkata    Banglore  18:55:00   
3  Multiple Carriers      2019-06-09     Delhi      Cochin  17:00:00   
4        Jet Airways      2019-03-12  Banglore   New Delhi  18:55:00   

  arrival_time  duration  total_stops              additional_info  price  
0     10:55:00      1075          1.0  In-flight meal not included  10539  
1     19:00:00      1380          2.0                      No info  14300  
2     10:05:00       910          1.0  In-flight meal not included   8529  
3     01:30:00       510          1.0                      No info   7408  
4     10:25:00       930          1.0                      No info  13817

In [39]:
export_data(X_test, y_test, "test")

airline date_of_journey    source destination  dep_time  \
0  Multiple Carriers      2019-06-09     Delhi      Cochin  13:00:00   
1  Multiple Carriers      2019-05-21     Delhi      Cochin  10:00:00   
2             Indigo      2019-04-27   Kolkata    Banglore  09:35:00   
3             Indigo      2019-05-24   Kolkata    Banglore  19:20:00   
4           Spicejet      2019-04-12  Banglore       Delhi  05:55:00   

  arrival_time  duration  total_stops additional_info  price  
0     21:00:00       480          1.0         No info   7408  
1     01:30:00       930          1.0         No info  14067  
2     12:20:00       165          0.0         No info   4804  
3     23:45:00       265          1.0         No info   3809  
4     08:35:00       160          0.0         No info   4423